### DISCUSSION TITLE
Why no PPP module in network library

### DISCUSSION THREAD

**Discussion by chefkurt** [micropython/micropython#10296] [RP2040 / Pico]
I wonder why there is no PPP module in the network library like it is in the ESP32 port where you can use:

```python
import network
network.PPP()
```

Is there a way to compile this into the firmware or is there a plan to do so? Could really use this for connecting a GSM module like Wi-Fi.

**Comment by jimmo** [comment]
@chefkurt Sorry about the delay responding.

The simple answer is that nobody has implemented it yet. On ESP32 it was easy because the Espressif SDK provides PPP support.

Like you say this would be very useful for implementing cellular support, and contributions would be very welcome!

**Reply by chefkurt** [reply]
@jimmo 
No worries about the delay. I'd like to see this feature getting added in the future. My knowledge in C is unfortunately very limited otherwise I could help to contribute. As it is already implemented in the ESP32 branch there must be an easy way to port this to the RP2040 right? Maybe somebody can give me some help to do so.

**Reply by jimmo** [reply]
Unfortunately the ESP32 implementation isn't going to be terribly useful as it is highly integrated with the IDF.

I think the place to start is to see how the SLIP implementation of our LWIP binding works, PPP will be similar. https://github.com/micropython/micropython/blob/master/extmod/modlwip.c#L91

**Reply by SirWintour** [reply]
I am not great at the backbones of network communications, so I have a few questions.

I noticed that the "[network_ppp.c](https://github.com/micropython/micropython/blob/master/ports/esp32/network_ppp.c)" in the esp32 port, just calls functions from the the "lwip/netif/ppp" in the ESP-IDF.

1. The pico-sdk has lwip/netif/ppp in it, can't we just point to that when building to the RP2040, or is it not that simple?
2. Is there complications with using the network, like how urequests just work on the ESP after connecting?
    2.1. Do we need to implement the routing manually?

**Reply by th3w** [reply]
As to point 1:
At first glance esp32/network_ppp.c  uses calls to RTOS functions such as xTaskCreatePinnedToCore, xTaskNotifyGive.
So it doesn't look like you can just link this code to the pico sdk version of LWIP

**Reply by victorallume** [reply]
Just wondering if there has been any movement on implementing PPP for non-ESP32 devices?

**Reply by ukrolelo** [reply]
+1 its very important ;(

**Comment by victorallume** [comment]
https://github.com/micropython/micropython/pull/14461
